src/models/filter.rs

```diff
@@ -193,68 +208,73 @@ impl SourceCodeUnit {
     &self, filter: Filter, node: Node, rule_store: &mut RuleStore,
     substitutions: &HashMap<String, String>,
   ) -> bool {
-    // This ensures that the below while loop considers the current node too when checking for filters.
-    // It does not make sense to check for filter if current node is a "leaf" node.
-    let mut initial_node = node;
+    let mut node_to_check = node;
+    let instantiated_filter = filter.instantiate(substitutions);
 
-    // No enclosing node is provided
-    if filter.enclosing_node().get_query().as_str() == DEFAULT_ENCLOSING_QUERY {
-      // Get the enclosing node matching the pattern specified in the filter (`filter.enclosing_node`)
-      self._check_current_node(filter, rule_store, substitutions, initial_node)
-    } else {
+    // If there is a not_enclosing_node is provided
+    let query = instantiated_filter.not_enclosing_node();
```

Comment:
are we saying that we can provide `enclosing` and `not-enclosing` patterns ?

So we are supporting 
optional[enclosing node] = Some(primary_match)
optional[not-enclosing node] = None
optional[contains] = None
optiona[atleast] = None (defaults to 1, when contains is provided)
optional[atmost] = None (defaults to MAX, when contains is provided) 
optional[not-contains] = None


And there is this data flow between enclosing -> contains -> atleast/atmost 


Also I am assuming in case of `not-enclosing` if we provide `contains` we will check it for the primary match